scripts/kconfig/kconfig.py

```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env python3
+# Modified from: https://github.com/ulfalizer/Kconfiglib/blob/master/examples/merge_config.py
+from kconfiglib import Kconfig, Symbol, BOOL, STRING, TRISTATE, TRI_TO_STR
```

Comment:
Just to keep things tidy and reviewable since this is not minimal, I think putting kconfiglib.py in ext/ (one patch to add the upstream copy, another with your changes) would be appropriate.

You can explicitly add `os.path.join(os.path.abspath(os.environ['ZEPHYR_BASE']), 'ext', 'subdir', 'containing', 'kconfiglib')` to sys.path before performing the import to ensure it's visible as done here:

https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/conf.py#L21

Reply:
I am not sure about that because `ext/` is supposed to hold *unmodified copies* of external repos. Instead I suggest we fix our Kconfig structure and then use `pip` to install Kconfiglib at a later date.

Reply:
@nashif can you weigh in?

Reply:
ext/ is for stuff that ends up in the binary, external components and middleware, we never put original  kconfig in ext/ and we also never put kbuild, kconfig is part of the build and configuration system and IMO should be in scripts/ just like it used to be before.

Reply:
and the other thing, we want to get rid of this eventually anyways and use the upstream version, so it will go away sooner than later.

Reply:
> ext/ is supposed to hold unmodified copies of external repos

That's not correct across the board. For example, STM32 HALs have received patches, mostly to fix upstream bugs:

0329450088fdb1ff9c6e31011496b907283cd950
9f420955d9a3cca4573a2a82d36e0e7b92cdcf3b

Looks like Nordic HALs are altered as well:

fc78ddf7d3a20d99669e49c0aba9ea1c9b8c702f
71c3d340da8cc43afb019d31756b818298ccddb9

The policy has been as I described in my experience. In the ST case, there are even README files tracking the patch list. I thought this was officially documented but can't find it right now

Reply:
Commit template for external components:
http://docs.zephyrproject.org/contribute/contribute_guidelines.html#identifying-contribution-origin

Requirement that non-Apache lives in ext/:
http://docs.zephyrproject.org/contribute/contribute_non-apache.html

Still can't find documentation about splitting up upstream import from Zephyr patches, but I'm sure that's generally what's expected

Reply:
> ext/ is for stuff that ends up in the binary, external components and middleware, we never put original kconfig in ext/ and we also never put kbuild, kconfig is part of the build and configuration system and IMO should be in scripts/ just like it used to be before.

Okay. Should I send patches to the documentation to make this more clear?

Reply:
In general you raise good points @mbolivar, but I still think `kconfiglib.py` should live in scripts/kconfig.

> That's not correct across the board. 

You are right, and the Nordic ones are definitely incorrect oversights that will be fixed soon once we replace what we have now with `nrfx`. 

> Commit template for external components:

Yes, I did not follow that strictly because I assumed that was only valid for `ext/` :) bit of a catch-22 here.

> Requirement that non-Apache lives in ext/:

I think that this applies only to firmware, i.e. source code that once compiled ends up in the image to flash.

Reply:
FWIW @mbolivar we even have GPL code in `scripts/kconfig`! https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/kconfig/gconf.c#L5

Reply:
kconfiglib.py ALREADY exists in the tree, this is just an update to an existing library we already have. We did not remove the old one because there is another script that depend on it for generating docs, but this should be done in parallel and be ready before this PR is merged. I started looking at that.

Reply:
OK OK, my mistake :).

I'm just going to send a docs patch to try to clear this up, we can discuss that there